backends/tc/ebpfCodeGen.cpp

```diff
@@ -0,0 +1,1481 @@
+    builder->blockStart();
+    builder->emitIndent();
+
+    pipeline->parser->headerType->declare(builder, "cpumap_hdr", false);
+    builder->endOfStatement(true);
+    builder->emitIndent();
+    auto user_md_type = pipeline->typeMap->getType(pipeline->control->user_metadata);
+    BUG_CHECK(user_md_type != nullptr, "cannot declare user metadata");
+    auto userMetadataType = EBPF::EBPFTypeFactory::instance->create(user_md_type);
+    userMetadataType->declare(builder, "cpumap_usermeta", false);
+    builder->endOfStatement(true);
+
+    // additional field to avoid compiler errors when both headers and user_metadata are empty.
+    builder->emitIndent();
+    builder->append("__u8 __hook");
+    builder->endOfStatement(true);
+
+    builder->blockEnd(false);
+    builder->endOfStatement(true);
+    builder->newline();
+}
+
+void PNAEbpfGenerator::emitPipelineInstances(EBPF::CodeBuilder *builder) const {
+    pipeline->parser->emitValueSetInstances(builder);
+    pipeline->deparser->emitDigestInstances(builder);
+
+    builder->target->emitTableDecl(builder, "hdr_md_cpumap", EBPF::TablePerCPUArray, "u32",
+                                   "struct hdr_md", 2);
+}
+
+// =====================PNAArchTC=============================
+void PNAArchTC::emit(EBPF::CodeBuilder *builder) const {
+    /**
+     * How the structure of a single C program for PSA should look like?
+     * 1. Automatically generated comment
+     * 2. Includes
+     * 3. Macro definitions (it's called "preamble")
+     * 4. Headers, structs, types
+     * 5. BPF map definitions.
+     * 6. XDP helper program.
+     * 7. TC Pipeline program.
+     */
+
+    // 1. Automatically generated comment.
+    // Note we use inherited function from EBPFProgram.
+    xdp->emitGeneratedComment(builder);
+
+    /*
+     * 2. Includes.
+     */
+    builder->target->emitIncludes(builder);
+    emitPNAIncludes(builder);
+
+    /*
+     * 3. Macro definitions (it's called "preamble")
+     */
+    emitPreamble(builder);
+
+    /*
+     * 4. Headers, structs, types
+     */
+    emitInternalStructures(builder);
+    emitTypes(builder);
+    emitGlobalHeadersMetadata(builder);
+
+    /*
+     * 5. BPF map definitions.
+     */
+    emitInstances(builder);
+
+    /*
+     * 6. XDP helper program.
+     */
+    xdp->emit(builder);
+
+    /*
+     * 7. TC Pipeline program.
+     */
+    pipeline->emit(builder);
+
+    builder->target->emitLicense(builder, pipeline->license);
+}
+
+void PNAArchTC::emitInstances(EBPF::CodeBuilder *builder) const {
+    builder->appendLine("REGISTER_START()");
+
+    if (options.xdp2tcMode == XDP2TC_CPUMAP) {
+        builder->target->emitTableDecl(builder, "xdp2tc_cpumap", EBPF::TablePerCPUArray, "u32",
+                                       "u16", 1);
+    }
+
+    emitPipelineInstances(builder);
+    builder->appendLine("REGISTER_END()");
+    builder->newline();
+}
+
+void PNAArchTC::emitInitializerSection(EBPF::CodeBuilder *builder) const {
```

Comment:
P4TC doesn't use BPF maps - `map-initializer` is not required. Please remove

Reply:
Removed